### PR TITLE
Rank Appts

### PR DESCRIPTION
**Pull request by kat-hall123**
## Description
- This PR takes the filtered appointments and ranks them in an order that best reduces gap time in the provider's schedule. Specifically, it looks at the gap time before and after each potential available appointment and ranks the smallest value appointment first. If there are no existing booked appointments, it simply returns appointments in an order that matches a provider's preference of earlier or later appointments.
- There were also slight formatting changes to make the code better readable (i.e. using startDateTime and no floating numbers)
- What's Next: Starting Technical Challenge 2 Fuzzy Search

## Milestones
- Technical Challenge (Smart Scheduling): Complete

## Test Plan
- First shows how no appointments existing just ranks appointments in either descending or ascending order
- Then shows when appointments are booked (9:30 and 11:20), it ranks the correct order
- 10 - 10:30 is first because it's right after 9:30 - 10
- 10:45 - 11:15 is second because it only has a 5 minute gap before the booked appointment at 11:20
- 10:15 is better than 10:30 because the provider prefers earlier appointments
- 11:00 was hidden from available because 11 - 11:30 would have occurred during the 11:20 appointment 

https://github.com/user-attachments/assets/982b192c-ccd4-4f06-8a24-f76803740b27

### CODE REVIEW

File: frontend/capstone-frontend/src/DashComponents/ProviderPageClientView.jsx [hunk 1]
```diff
@@ -172,18 +185,65 @@ function ProviderPageClientView() {
...
+                .filter(booked => new Date(booked.endDateTime) <= appointmentStart) // Find booked appointments occurring BEFORE potential appointment
+                .sort((a, b) => new Date(b.endDateTime) - new Date(a.endDateTime)) // Sort by putting closest before potential appointment in first position 
+            
+            const mostPrevious = appointmentsBefore[0]  // Get the closest appointment before 
+
+            const gapBefore = mostPrevious ? appointmentStart - new Date(mostPrevious.endDateTime) : 0 // If appointment before exists, find gap before potential appointment; if not, use 0
+
+            /* Handle Gap After */
+            const appointmentsAfter = bookedAppointments
+                .filter(booked => new Date(booked.startDateTime) >= appointmentEnd) // Find booked appointments occurring AFTER potential appointment
```
Comment (maurenej) [inline frontend/capstone-frontend/src/DashComponents/ProviderPageClientView.jsx:219]: Just checking--do you currently have a check to ensure the provider can't put up two appointment slots that start at the same time? I _think_ you do but want to make sure it's not possible that you have two appointments with the same start and end times since that appointment would be present in both `appointmentsBefore` and `appointmentsAfter`.

File: frontend/capstone-frontend/src/DashComponents/ProviderPageClientView.jsx [hunk 2]
```diff
@@ -172,18 +185,65 @@ function ProviderPageClientView() {
...
+        }).sort((a, b) => {
+            if(a.bothGaps[0] !== b.bothGaps[0]) {
+                return a.bothGaps[0] - b.bothGaps[0] // Ascending order with smallest gap time first
+            }
+            if(a.bothGaps[1] !== b.bothGaps[1]) {
+                return a.bothGaps[1] - b.bothGaps[1]
+            }
+
+            // If both equal gap times, rank by provider's preference of earlier/later appts
+            return prefersEarly ? a.appointmentStart - b.appointmentStart : b.appointmentStart - a.appointmentStart
```
Comment (maurenej) [inline frontend/capstone-frontend/src/DashComponents/ProviderPageClientView.jsx:244]: Really like the edge case handling here!

**Review by maurenej — APPROVED** [review]
Awesome job on implementing your algorithm! The recommendation feature also looks great in your test plan :) Just left a quick question but looks great!

File: frontend/capstone-frontend/src/DashComponents/ProviderPageClientView.jsx (hunk 1 above)
Comment (kat-hall123) [reply frontend/capstone-frontend/src/DashComponents/ProviderPageClientView.jsx:219]: I have a check for providers trying to add an existing availability with same start time as well as trying to start it in the middle of an existing appointment. Is this what you mean?
https://github.com/user-attachments/assets/bd5eb61a-b303-4374-9d14-e53ef1d0147c

Comment (maurenej) [reply frontend/capstone-frontend/src/DashComponents/ProviderPageClientView.jsx:219]: Oh perfect! Yup, this is what I was looking for :)
